File: lab_1/task_2.py
import math

class Task1:
    def __init__(self):
        self.x0 = 81
        self.x = 80.5
        self.dx = self.x - self.x0
    
    def func(self, x):
        return (math.sqrt(x) + 1) * (math.sqrt(x) + 1)
    
    def df_dx(self, x):
        return 1 + 1 / math.sqrt(x)
    
    def a(self):
        return self.func(self.x0) + self.df_dx(self.x0) * self.dx
    
    def absolute_error(self):
        return abs(self.func(self.x) - self.a())
    
    def relative_error(self):
        return (self.absolute_error() / self.func(self.x)) * 100


class Task2:
    def __init__(self):
        self.x0 = 1.0
        self.y0 = 1.0
        self.x = 1.02
        self.y = 0.97
        self.dx = self.x - self.x0
        self.dy = self.y - self.y0
    
    def func(self, x, y):
        return math.log(x + y)
    
    def df_dx(self, x, y):
        return 1.0 / (x + y)
    
    def df_dy(self, x, y):
        return 1.0 / (x + y)
    
    def a(self):
        return self.func(self.x0, self.y0) + self.df_dx(self.x0, self.y0) * self.dx + self.df_dy(self.x0, self.y0) * self.dy
    
    def absolute_error(self):
        return abs(self.func(self.x, self.y) - self.a())
    
    def relative_error(self):
        return (self.absolute_error() / self.func(self.x, self.y)) * 100


if __name__ == "__main__":
    task1 = Task1()
    print("task 1:")
    print(f"x0 = {task1.x0}")
    print(f"x = {task1.x}")
    print(f"dx = {task1.dx}")
    print(f"A = {task1.func(80.5):.15f}")
    print(f"a = {task1.a():.15f}")
    print(f"Absolute error: {task1.absolute_error():.6f}")
    print(f"Relative error: {task1.relative_error():.6f}%\n")
    
    task2 = Task2()
    print("task 2:")
    print(f"x0 = {task2.x0}, y0 = {task2.y0}")
    print(f"x = {task2.x}, y = {task2.y}")
    print(f"dx = {task2.dx}, dy = {task2.dy}")
    print(f"A = {task2.func(task2.x, task2.y):.15f}")
    print(f"a = {task2.a():.15f}")
    print(f"Absolute error: {task2.absolute_error():.6f}")
    print(f"Relative error: {task2.relative_error():.6f}%")

File: lab_1/task_2.java
public class task_2 {
    
    static class Task1 {
        double x0 = 81.0;
        double x = 80.5;
        double dx = x - x0;
        
        double func(double x) {
            return (Math.sqrt(x) + 1) * (Math.sqrt(x) + 1);
        }
        
        double df_dx(double x) {
            return 1 + 1 / Math.sqrt(x);
        }
        
        double a() {
            return func(x0) + df_dx(x0) * dx;
        }
        
        double absoluteError() {
            return Math.abs(func(x) - a());
        }
        
        double relativeError() {
            return (absoluteError() / func(x)) * 100;
        }
    }
    
    static class Task2 {
        double x0 = 1.0, y0 = 1.0;
        double x = 1.02, y = 0.97;
        double dx = x - x0, dy = y - y0;
        
        double func(double x, double y) {
            return Math.log(x + y);
        }
        
        double df_dx(double x, double y) {
            return 1.0 / (x + y);
        }
        
        double df_dy(double x, double y) {
            return 1.0 / (x + y);
        }
        
        double a() {
            return func(x0, y0) + df_dx(x0, y0) * dx + df_dy(x0, y0) * dy;
        }
        
        double absoluteError() {
            return Math.abs(func(x, y) - a());
        }
        
        double relativeError() {
            return (absoluteError() / func(x, y)) * 100;
        }
    }
    
    public static void main(String[] args) {
        Task1 task1 = new Task1();
        System.out.println("task 1:");
        System.out.println("x0 = " + task1.x0);
        System.out.println("x = " + task1.x);
        System.out.println("dx = " + task1.dx);
        System.out.printf("A = %.15f%n", task1.func(80.5));
        System.out.printf("a = %.15f%n", task1.a());
        System.out.printf("Absolute error: %.6f%n", task1.absoluteError());
        System.out.printf("Relative error: %.6f%%%n%n", task1.relativeError());
        
        Task2 task2 = new Task2();
        System.out.println("task 2:");
        System.out.println("x0 = " + task2.x0 + ", y0 = " + task2.y0);
        System.out.println("x = " + task2.x + ", y = " + task2.y);
        System.out.println("dx = " + task2.dx + ", dy = " + task2.dy);
        System.out.printf("A = %.15f%n", task2.func(task2.x, task2.y));
        System.out.printf("a = %.15f%n", task2.a());
        System.out.printf("Absolute error: %.6f%n", task2.absoluteError());
        System.out.printf("Relative error: %.6f%%%n", task2.relativeError());
    }
}
